x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/planner/QueryTranslator.java

```diff
@@ -461,23 +456,7 @@ protected QueryTranslation asQuery(In in, boolean onAggs, TranslatorHandler hand
                 aggFilter = new AggFilter(id(in.value()), in.asScript());
             }
             else {
-                Query q = null;
-                if (in.value() instanceof FieldAttribute) {
-                    // equality should always be against an exact match (which is important for strings)
-                    FieldAttribute fa = (FieldAttribute) in.value();
-                    List<Expression> list = in.list();
-                    // TODO: this needs to be handled inside the optimizer
-                    list.removeIf(e -> DataTypes.isNull(e.dataType()));
-                    DataType dt = list.get(0).dataType();
-                    Set<Object> set = new LinkedHashSet<>(CollectionUtils.mapSize(list.size()));
-                    for (Expression e : list) {
-                        set.add(SqlDataTypeConverter.convert(e.fold(), dt));
```

Comment:
One thing to note @costin is that `SqlDataTypeConverter` isn't available in QL,  so this is now just `valueOf`. Looks like SQL tests are still passing though
https://github.com/elastic/elasticsearch/blob/d6757d01e376a233c45537814452b43d7ff84b94/x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/planner/ExpressionTranslators.java#L345

Reply:
This can addressed by adding a `convertTo()` method which by default uses `DataTypeConverter` inside QL and `SqlDataTypeConverter` inside SQL.
That said, if the tests are passing across the board I would leave this as is to avoid doing any implicit conversion.
Potentially add a comment mentioning that before there was a conversion happening to help track this down in case a weird bug occurs in the future.

Reply:
Looks like the failing tests are from the SQL tests so the solution would be to keep the extend `converTo` in the SQL `ExpressionTranslator` and use `SqlDataTypeConverter` there as before.

Reply:
Looking some more, I see this method that could also work. Could just make it `public` instead of `protected` and use this:
https://github.com/elastic/elasticsearch/blob/7425042f86f66df8c207c5e96f9b9848bda2b4c3/x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/expression/predicate/operator/comparison/In.java#L106-L112

And since In is subclassed in SQL, this has an override for SqlDataTypeConverter
https://github.com/elastic/elasticsearch/blob/7425042f86f66df8c207c5e96f9b9848bda2b4c3/x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/expression/predicate/operator/comparison/In.java#L106-L112

Reply:
I think you're suggesting to use the `foldAndConvertList` from `In` outside of it.
We're trying to prevent that since when we expanded the types, it was discovered that there were quite a number of implicit conversions which are simply bad practices since they appear "magic" and allow lenient behavior.
We actually had a `Foldables` class that did that but it wasn't used anywhere outside `In` hence why it got moved there.
If we are to make it public and end up using it, we're back where we started.

This method might fix the current sets of tests however I expect more `ExpressionTranslation` rule in the future to require the `valueOf` method and thus delegation to the appropriate converter for their respective projects.